Flaky test: TestPreinteropFaultProofs_VariedBlockTimes

**Claude:** Reporting flaky test.

## Flaky Test

- **Test:** `TestPreinteropFaultProofs_VariedBlockTimes` (and `_FasterChainB` variant)
- **Package:** `github.com/ethereum-optimism/optimism/op-acceptance-tests/tests/isthmus/preinterop`
- **Status:** Currently skipped with `t.Skip("Skipping flaky varied block time fault proof test")`
- **Related:** ethereum-optimism/optimism#19010

## Context

The test configures two interop chains with different block times (1s and 2s), freezes batchers, captures L1 heads at different batch-availability points, then runs super-root transition test cases through both kona interop FPP and challenger trace provider.

The test was skipped due to observed flakiness. Investigating root cause.

**Claude:** Reopening — same test flaked again with a different root cause.

**CI:** https://app.circleci.com/pipelines/github/ethereum-optimism/optimism/124540/workflows/b9138843-de0a-44f3-939c-92ebd0977d15/jobs/4991666/tests
**Job:** memory-all-opn-op-geth (run #4991666, 2026-05-06)

**Failure:**
```
superfaultproofs.go:78
Error: "1778102404" is not less than or equal to "1778102403"
Messages: chain 901 head 1778102404 is already past target 1778102403
--- FAIL: TestPreinteropFaultProofs_VariedBlockTimes (88.63s)
```

The assertion is in `advanceUnsafeToTimestamp` (superfaultproofs.go:78), failing with `target = startTimestamp` (`endTimestamp - 1`).

`endTimestamp` is computed by `nextTimestampAfterSafeHeads` at line 175, which uses `LocalSafeL2.Number + 1` per chain. After `freezeChains`, the unsafe head can be several blocks ahead of local safe (batcher lag); `startTimestamp = endTimestamp - 1` can therefore be less than the current unsafe head's timestamp on the faster chain (901, blockTime=1s), tripping the assertion. Whether this trips depends on how far unsafe ran ahead of safe before freezeChains stopped batchers — i.e. timing.

Investigating fix.

--- a/op-acceptance-tests/tests/superfaultproofs/superfaultproofs.go
+++ b/op-acceptance-tests/tests/superfaultproofs/superfaultproofs.go
@@ -45,20 +45,30 @@ type chain struct {
 	Batcher *dsl.L2Batcher
 }
 
-// freezeChains takes exclusive control of L2 block production: stops every
-// chain's real sequencer and batcher, and waits for both safety levels to
-// stall. After this returns no real-sequencer or batcher activity advances
-// the chain — block production is the caller's via TestSequencer.SequenceBlock.
+// freezeChains takes exclusive control of L2 block production. After this
+// returns: real sequencers are stopped, batchers are stopped, and on every
+// chain LocalSafe == LocalUnsafe. From here, only the caller advances the
+// chain via TestSequencer.SequenceBlock.
+//
+// Order matters: stopping the batcher before LocalSafe catches up to
+// LocalUnsafe leaves safe behind unsafe, which breaks downstream code that
+// expects "the head" to be unambiguous (e.g. nextTimestampAfterSafeHeads).
+// So we stop sequencers first, wait for unsafe to stall, then let the still-
+// running batchers drain the remaining unsafe blocks to L1 until safe catches
+// up, and only then stop the batchers.
 func freezeChains(chains []*chain) {
 	for _, c := range chains {
 		c.CLNode.StopSequencer()
 	}
 	for _, c := range chains {
-		c.Batcher.Stop()
+		c.CLNode.WaitForStall(types.LocalUnsafe)
 	}
 	for _, c := range chains {
-		c.CLNode.WaitForStall(types.LocalUnsafe)
-		c.CLNode.WaitForStall(types.LocalSafe)
+		unsafeNumber := c.CLNode.HeadBlockRef(types.LocalUnsafe).Number
+		c.CLNode.Reached(types.LocalSafe, unsafeNumber, 30)
+	}
+	for _, c := range chains {
+		c.Batcher.Stop()
 	}
 }
 

